thirdparty/recpinset/recpinset.go

```diff
@@ -0,0 +1,130 @@
+package recpinset
+
+import (
+	cid "gx/ipfs/QmapdYm1b22Frv3k17fqrBYTFRxwiaVJkB299Mfn33edeB/go-cid"
+)
+
+// Set stores a set of recursive pins (Cid,MaxDepth tuples), indexed by Cid.
+// A MaxDepth of -1 means fully no-limit recursive pin.
+type Set struct {
+	set map[string]int
+}
+
+// RecPin represents a recursive Pin, that is, a CID and a MaxDepth that
+// limits the depth to which the CID is pinned. A negative MaxDepth represents
+// a unlimited-depth pin.
+type RecPin struct {
+	Cid      *cid.Cid
+	MaxDepth int
+}
+
+// New initializes and returns a new Set.
+func New() *Set {
+	return &Set{set: make(map[string]int)}
+}
+
+// Add puts a Cid and the maxDepth associated to it in the Set.
+func (s *Set) Add(c *cid.Cid, maxDepth int) {
+	s.set[string(c.Bytes())] = maxDepth
+}
+
+// Has returns if the Set contains a given Cid.
+func (s *Set) Has(c *cid.Cid) bool {
+	_, ok := s.set[string(c.Bytes())]
+	return ok
+}
+
+// Get returns the RecPin associated to the given Cid.
+func (s *Set) Get(c *cid.Cid) (*RecPin, bool) {
+	md, ok := s.set[string(c.Bytes())]
+	if !ok {
+		return nil, false
+	}
+
+	return &RecPin{c, md}, ok
+}
+
+// MaxDepth returns the MaxDepth associated to the given Cid.
+func (s *Set) MaxDepth(c *cid.Cid) (int, bool) {
+	md, ok := s.set[string(c.Bytes())]
+	return md, ok
+}
+
+// Remove deletes a Cid from the Set.
+func (s *Set) Remove(c *cid.Cid) {
+	delete(s.set, string(c.Bytes()))
+}
+
+// Len returns how many elements the Set has.
+func (s *Set) Len() int {
+	return len(s.set)
+}
+
+// Keys returns the Cids in the set.
+func (s *Set) Keys() []*cid.Cid {
+	out := make([]*cid.Cid, 0, len(s.set))
+	for k := range s.set {
+		c, _ := cid.Cast([]byte(k))
+		out = append(out, c)
+	}
+	return out
+}
+
+// RecPins returns all the Cid+MaxDepths stored in the set,
+// in a slice of RecPins.
+func (s *Set) RecPins() []*RecPin {
+	out := make([]*RecPin, 0, len(s.set))
+	for k, v := range s.set {
+		c, _ := cid.Cast([]byte(k))
+		out = append(out, &RecPin{c, v})
+	}
+	return out
+}
+
+// Visit adds a cid and maxDepth to the set if:
+// - it is not already in the set
+// - if it's in the set but the new maxDepth is greater than
+// the existing.
+// It returns true if the RecPin has been added.
+func (s *Set) Visit(c *cid.Cid, maxDepth int) bool {
+	curMaxDepth, ok := s.set[string(c.Bytes())]
+
+	if !ok || IsDeeper(maxDepth, curMaxDepth) {
```

Comment:
```
[ A ]
  |     \
[ B ]  [ C ]
 |          |
[ C ]  [ D ]
```

So what happens if we visit C in the first (left) tree? It seems like we would call visit on the second one, and pass `1,2` to `IsDeeper` which would return false, and cause us to not ever visit D.

Reply:
Sorry, but that's not correct.

Assuming you are traversing this graph with MaxDepth=2 at the beginning:

* The C on the first branch would be visited with maxDepth=0 `(c,0)`
* The C on the second branch would be visited with MaxDepth=1 (c,1)`

1>0, thus Visit will return `true`, thus the functions will keep traversing the path because the previously visited C had a lower depth limit than the new one. 

In this context `maxDepth` means, `the maximum depth of the tree below this CID`. Thus we always keep exploring if `IsDeeper()`. It does not mean `item's depth` as you seemed to assume.

Reply:
Ah, these variables could definitely use some better naming then. and maybe a comment. maybe `maxDepth` -> `curHeight` ?

Reply:
It is actually `maxDepth`. That is, it's the "max depth" at which we are planning on exploring this CID. When we *increase* the `maxDepth`, we explore the path again.